File: 14888-연산자 끼워넣기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main(){
    int N;

    cin>>N;

    vector<int> operands(N);
    vector<int> operators;

    for(int i=0; i<N; i++){
        cin>>operands[i];
    }

    for(int i=0; i<4; i++){
        int n;
        cin>> n;

        for(int j=0; j<n; j++){
            operators.push_back(i);
        }
    }


    int min_value = 1'000'000'000;
    int max_value = -1'000'000'000;

    do{
        int result = operands[0];
        int idx = 0;
        
        for(int i=0; i<N-1; i++){
            idx++;
            int op = operators[i];

            if (op == 0)
                result += operands[idx];
            else if(op == 1)
                result -= operands[idx];
            else if(op == 2)
                result *= operands[idx];
            else if(op == 3)
                result /= operands[idx];
        }

        min_value = min(min_value, result);
        max_value = max(max_value, result);

    }while(next_permutation(operators.begin(),operators.end()));

    cout<<max_value<<endl;
    cout<<min_value<<endl;
}

File: 14888-연산자 끼워넣기.py
N = int(input())
operands = list(map(int, input().split()))
num_of_operator = list(map(int, input().split()))


# 10P10 10! => 3628800

'''
operator
0 : +
1 : -
2 : *
3 : //
'''

def permutation(arr,n):
    result = []

    if n == 0:
        return [[]]

    for i, e in enumerate(arr):
        for p in permutation(arr[:i] + arr[i+1:], n-1):
            result += [[e] + p]
    
    return result

def caculate(operands,operators):
    result = operands[0]

    for i,j in zip(operands[1:],operators):
        if j == 0:
            result += i
        elif j == 1:
            result -= i
        elif j == 2:
            result *= i
        elif j == 3:
            if result < 0:
                result = -(-result // i)
            else:
                result //= i

    return result

flatten_operator = []

for i,n in enumerate(num_of_operator):
    for j in range(n):
        flatten_operator.append(i)

min_value = 1e9
max_value = -1e9

for operators in permutation(flatten_operator,N-1):
    result = caculate(operands,operators)

    max_value = min(result,max_value)
    min_value = min(result,min_value)

print(max_value)
print(min_value)